apps/aechannel/src/aesc_fsm_min_depth_watcher.erl

```diff
@@ -104,122 +243,699 @@ terminate(_Reason, _St) ->
+              , changed      => Changed
+              , channel      => Ch
+              , lock_period  => aesc_channels:lock_period(Ch)
+              , locked_until => aesc_channels:locked_until(Ch) },
+             C1}
+    end.
+
+get_channel(ChId, Hash, C) ->
+    cached_get({channel, Hash}, C, fun() -> get_channel_(ChId, Hash) end).
+
+get_channel_(ChId, Hash) ->
+    case aec_chain:get_block_state(Hash) of
+        {ok, Trees} ->
+            case aec_chain:get_channel(ChId, Trees) of
+                {ok, Ch} ->
+                    Ch;
+                {error, _} ->
+                    undefined
+            end;
+        error ->
+            undefined
+    end.
+
+cached_get(Key, C, Get) ->
+    case maps:find(Key, C) of
+        {ok, V} ->
+            {V, C};
+        error ->
+            get_and_cache(Get, Key, C)
+    end.
+
+get_and_cache(F, Key, C) when is_function(F, 0) ->
+    V = F(),
+    {V, C#{ Key => V }};
+get_and_cache(F, Key, C) when is_function(F, 1) ->
+    {V, C1} = F(C),
+    {V, C1#{ Key => V }}.
+
+log_tx(#{ tx_hash      := TxHash
+        , block_hash   := BlockHash
+        , block_origin := _Origin
+        , type         := _Type } = Info, #st{tx_log = TxLog} = St) ->
+    Key = {TxHash, BlockHash},
+    case aesc_window:keymember(Key, 1, TxLog) of
+        true ->
+            St;
+        false ->
+            TxLog1 = aesc_window:add({Key, Info}, TxLog),
+            St#st{tx_log = TxLog1}
+    end.
+
+get_latest_tx(Log) ->
+    {{{_TxHash, _BlockHash}, Tx}, _TxLog1} = aesc_window:pop(Log),
+    lager:debug("Tx = ~p", [Tx]),
+    Tx.
+
+%% Find recent txs. Two different stop conditions:
+%% - {any_after_block, BlockHash}: Stop as soon as a block is found with
+%%   relevant txs in it, or when BlockHash is reached.
+%% - {all_after_tx, TxHash}: Find all txs after TxHash. Note that we must
+%%   first verify that TxHash exists on chain.
+%% Find recent txs related to ChId since LastBlock (not including those in
+%% LastBlock) up until, and including those in, Hash.
+%% This is like aec_chain:get_transactions_between/2, except it ignores all
+%% txs other than channel txs related to ChId, groups by block, etc.
+%%
+get_txs_since(StopCond, Hash, ChId, C) ->
+    case verify_stop_cond(StopCond, C) of
+        {ok, C1} ->
+            lager:debug("StopCond = ~p", [StopCond]),
+            TxLog = maps:get(tx_log, C1),
+            {Found, C2} = get_txs_since(StopCond, Hash, ChId, C1, []),
+            TxLog1 =
+                lists:foldl(
+                  fun({BlockHash, Txs}, Acc) ->
+                          lists:foldl(
+                            fun(TxHash, Acc1) ->
+                                    aesc_window:add(
+                                      { {TxHash, BlockHash},
+                                        #{ tx_hash      => TxHash
+                                         , block_hash => BlockHash
+                                         , block_origin => chain } }, Acc1)
```

Comment:
The second element in this tuple that is the entry of the `aesc_window` data structure has no `type` key here.